File: gym_cdm2/torchModule.py
import argparse
import os
# workaround to unpickle olf model files
import sys
import platform
import numpy as np
import torch
import pdb # use pdb.set_trace() for debugging

sys.path.append(os.getcwd())
sys.path.append('gym_cdm')

import work.settings as settings

class EnvNormalize:
    """
    A non-vectorized wrapper by taesoo that normalizes the observations
    """
    def __init__(self, ob_rms, ret=True, clipob=10., cliprew=10., gamma=0.99, epsilon=1e-8 ):
        self.ob_rms = ob_rms
        self.clipob = clipob
        self.gamma = gamma
        self.epsilon = epsilon


    def process(self, obs):
        if self.ob_rms:
            obs = np.clip((obs - self.ob_rms.mean) / np.sqrt(self.ob_rms.var + self.epsilon), -self.clipob, self.clipob)
            return torch.from_numpy(obs).float()
        else:
            return obs

def loadAgent(spec_id, filename):
    if not (hasattr(settings, 'agents')):
        settings.agents={}

    ac, ob_rms= torch.load(filename, weights_only=False)
    vec_norm =EnvNormalize(ob_rms);
    agent={
            'actor_critic': ac, 
            'ob_rms': ob_rms,
            'vec_norm':vec_norm,
            }
    settings.agents[spec_id]=agent


def getAction(spec_id, obs, taction):
    agent=settings.agents[spec_id]
    actor_critic=agent['actor_critic']
    obs=agent['vec_norm'].process(obs.ref())
    base=actor_critic.base
    with torch.no_grad():
        actor_features=base.actor(obs)
        action=actor_critic.dist.fc_mean(actor_features)
        action=action.cpu().numpy().astype(np.float64)
    taction.setSize(len(action))
    taction.ref()[:]=action

File: gym_cdm2/torchModule.lua
EnvNormalize=LUAclass()
function EnvNormalize:__init(ob_rms, clipob, gamma, epsilon)
	assert(ob_rms)
	self.ob_rms=ob_rms
	self.clipob=clipob or 10.0
	self.gamma=gamma or 0.99
	self.epsilon=epsilon or 1e-8
end
function EnvNormalize:process(obs)
	local obs = np.clip((obs - self.ob_rms.mean) / np.sqrt(self.ob_rms.var + self.epsilon), -self.clipob, self.clipob)
	return obs
end

TorchAgent=LUAclass()

function TorchAgent:__init(filename, spec_id)
	self.spec_id=spec_id
	if not TorchModule then
		-- use python3 instead.
		python.F('gym_cdm2.torchModule', 'loadAgent', spec_id, filename)
		return
	end

	if true then -- always rebuild module. 
	--if not (os.isFileExist(filename..'.module')) then
		-- this script first calculates a hash to avoid repeated conversions.
		os.execute('python3 convertPTtoTorchScript.py --env-name='..spec_id)
	end
	if os.isFileExist(filename..'.ob_rms') then
		--print('loading '..filename..'.module')
		--self.module=TorchModule(filename..".module")
		print('loading '..filename..'.ptl')
		self.module=TorchModule(filename..".ptl")

		local ob_rms={}
		ob_rms.mean=vectorn()
		ob_rms.var=vectorn()
		do
			local file=util.BinaryFile()
			file:openRead(filename..'.ob_rms')
			file:unpack(ob_rms.mean)
			file:unpack(ob_rms.var)
			file:close()
		end

		self.vec_norm=EnvNormalize(ob_rms)
	end
end

function TorchAgent:reset(userInput)
	local obs=reset(0)
	return obs
end


function TorchAgent:getAction(obs)
	if not TorchModule then
		local action=vectorn()
		-- use python3 instead.
		python.F('gym_cdm2.torchModule', 'getAction', self.spec_id, obs, action)
		return action
	end
	if self.vec_norm then
		obs=self.vec_norm:process(obs)
	end
    local action
	if self.module then
		assert(self.vec_norm)
		action=self.module:forward(obs)
	else
		print('using zero action')
		action=CT.zeros(get_dim()(1))
	end
	return action
end
